File: CPP/binary_tree_inorder_traversal.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> result;
        stack<TreeNode*> tn_stack;
        while (root != NULL || !tn_stack.empty()) {
            while (root != NULL) {
                tn_stack.push(root);
                root = root->left;
            }
            if (!tn_stack.empty()) {
                root = tn_stack.top();
                tn_stack.pop();
                result.push_back(root->val);
                root = root->right;
            }
          
        }
        return result;
    }
};

File: JAVA/BinaryTreeInorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new LinkedList<Integer>();
        Stack<TreeNode> nodes = new Stack<TreeNode>();
        while (root != null || !nodes.isEmpty()) {
            while(root != null) {
                nodes.push(root);
                root = root.left;
            }
            root = nodes.pop();
            result.add(root.val);
            root = root.right;
        }
        return result;
    }
}
